src/file/writer.rs

```diff
@@ -15,16 +15,290 @@
+}
+
+impl SerializedRowGroupWriter {
+  pub fn new(
+    schema_descr: SchemaDescPtr,
+    properties: WriterPropertiesPtr,
+    file: &File
+  ) -> Self {
+    let num_columns = schema_descr.num_columns();
+    Self {
+      descr: schema_descr,
+      props: properties,
+      file: file.try_clone().unwrap(),
+      total_rows_written: None,
+      total_bytes_written: 0,
+      column_index: 0,
+      current_column_writer: None,
+      row_group_metadata: None,
+      column_chunks: Vec::with_capacity(num_columns)
+    }
+  }
+
+  /// Checks and finalises current column writer.
+  fn finalise_column_writer(&mut self) -> Result<()> {
+    if let Some(ref mut writer) = self.current_column_writer {
+      let (bytes_written, rows_written, metadata) = match writer {
+        ColumnWriter::BoolColumnWriter(typed) => typed.close()?,
+        ColumnWriter::Int32ColumnWriter(typed) => typed.close()?,
+        ColumnWriter::Int64ColumnWriter(typed) => typed.close()?,
+        ColumnWriter::Int96ColumnWriter(typed) => typed.close()?,
+        ColumnWriter::FloatColumnWriter(typed) => typed.close()?,
+        ColumnWriter::DoubleColumnWriter(typed) => typed.close()?,
+        ColumnWriter::ByteArrayColumnWriter(typed) => typed.close()?,
+        ColumnWriter::FixedLenByteArrayColumnWriter(typed) => typed.close()?
+      };
+
+      // Update row group writer metrics
+      self.total_bytes_written += bytes_written;
+      self.column_chunks.push(metadata);
+      if let Some(rows) = self.total_rows_written {
+        if rows != rows_written {
+          return Err(general_err!(
+            "Incorrect number of rows, expected {} != {} rows",
+            rows,
+            rows_written
+          ));
+        }
+      } else {
+        self.total_rows_written = Some(rows_written);
+      }
+    }
+    Ok(())
+  }
+}
+
+impl RowGroupWriter for SerializedRowGroupWriter {
+  #[inline]
+  fn num_columns(&self) -> usize {
+    self.descr.num_columns()
+  }
+
+  #[inline]
+  fn has_next_column(&self) -> bool {
+    self.column_index < self.num_columns() && self.row_group_metadata.is_none()
+  }
+
+  #[inline]
+  fn next_column(&mut self) -> Result<&mut ColumnWriter> {
+    if self.row_group_metadata.is_some() {
+      return Err(general_err!("Row group writer is closed"));
+    }
+
+    self.finalise_column_writer()?;
+    if self.column_index < self.num_columns() {
+      let sink = FileSink::new(&self.file);
+      let page_writer = Box::new(SerializedPageWriter::new(sink));
+      let column_writer = get_column_writer(
+        self.descr.column(self.column_index),
+        self.props.clone(),
+        page_writer
+      );
+
+      self.column_index += 1;
+
+      self.current_column_writer = Some(column_writer);
+      Ok(self.current_column_writer.as_mut().unwrap())
+    } else {
+      Err(general_err!("No more column writers left"))
+    }
+  }
+
+  #[inline]
+  fn close(&mut self) -> Result<RowGroupMetaDataPtr> {
```

Comment:
what if user calls `close()` before iterating over all columns. Will that cause some error?

Reply:
Yes, it will raise an error, because there will be mismatch between expected columns (must be greater than 0) in the schema and actual number of columns written.